s3-migrate/migrate.py

```diff
@@ -0,0 +1,248 @@
+
+    @classmethod
+    def start_pool(cls, queue):
+        """Run threads on the data provided in the queue. Return a list of
+threads to pass to `finish_pool` later."""
+        threads = []
+        for _ in range(20):
+            thread = cls(queue)
+            threads.append(thread)
+            thread.start()
+        return threads
+
+    @staticmethod
+    def finish_pool(queue, threads):
+        """Wait until queue is empty and all threads have finished
+processing."""
+        # block until all tasks are done
+        queue.join()
+
+        # stop workers
+        for thread in threads:
+            thread.join()
+
+
+class MogileFile():
+    """Represents a file stored in mogile."""
+
+    def __init__(self, fid: int, dkey: str, length: int):
+        self.length = length
+        self.fid = fid
+        self.dkey = dkey
+        (self.sha1sum, self.orig_bucket) = dkey.split('-', 1)
+
+    @classmethod
+    def parse_row(cls, row: dict):
+        """Factory method to take an row as returned from the database and
+return a MogileFile."""
+        # Sanity check, we only use one "domain" and one class
+        assert row['dmid'] == 1, "Bad domain"
+        assert row['classid'] == 0, "Bad class"
+        return cls(dkey=row['dkey'],
+                   fid=row['fid'],
+                   length=row['length'])
+
+    def exists_in_bucket(self, client, bucket, key):
+        """Return True if object exists in the bucket."""
+        obj = client.Object(bucket, key)
+        try:
+            obj.load()
+            if obj.content_length != self.length:
+                return False
+        except ClientError as ex:
+            if ex.response['Error']['Code'] == "404":
+                return False
+            raise
+        return True
+
+    def make_intermediary_key(self):
+        """Return the key to use for the intermediary, mogile-based storage of
+this object in S3."""
+        padded = "{:010d}".format(self.fid)
+        return "{first}/{second}/{third}/{padded}.fid".format(
+            first=padded[0:1],
+            second=padded[1:4],
+            third=padded[4:7],
+            padded=padded)
+
+    def make_contentrepo_key(self):
+        """Return the key to use for the final storage of this object in S3."""
+        return self.sha1sum
+
+    def intermediary_exists_in_bucket(self, client, bucket):
+        """Return True if the intermediary (mogile-style key) object exists in
+the bucket."""
+        return self.exists_in_bucket(client, bucket,
+                                     self.make_intermediary_key())
+
+    def contentrepo_exists_in_bucket(self, client, bucket):
+        """Return True if the final contentrepo object exists in the bucket."""
+        return self.exists_in_bucket(client, bucket,
+                                     self.make_contentrepo_key())
+
+    def get_mogile_url(self, client):
+        """Get a URL from mogile that we can use to access this file."""
+        return random.choice(
```

Comment:
is this random used to distribute requests across the mogile nodes?

Reply:
Yep!